api/inc/rpc_gateway.h

```diff
@@ -0,0 +1,224 @@
+
+#define _UVISOR_BOX_RPC_GATEWAY_ARG_CHECK_2(gw_name, p0_type, p1_type) \
+    UVISOR_STATIC_ASSERT(sizeof(p0_type) <= sizeof(uint32_t), gw_name ## _param_0_too_big); \
+    UVISOR_STATIC_ASSERT(sizeof(p1_type) <= sizeof(uint32_t), gw_name ## _param_1_too_big);
+
+#define _UVISOR_BOX_RPC_GATEWAY_ARG_CHECK_3(gw_name, p0_type, p1_type, p2_type) \
+    UVISOR_STATIC_ASSERT(sizeof(p0_type) <= sizeof(uint32_t), gw_name ## _param_0_too_big); \
+    UVISOR_STATIC_ASSERT(sizeof(p1_type) <= sizeof(uint32_t), gw_name ## _param_1_too_big); \
+    UVISOR_STATIC_ASSERT(sizeof(p2_type) <= sizeof(uint32_t), gw_name ## _param_2_too_big);
+
+#define _UVISOR_BOX_RPC_GATEWAY_ARG_CHECK_4(gw_name, p0_type, p1_type, p2_type, p3_type) \
+    UVISOR_STATIC_ASSERT(sizeof(p0_type) <= sizeof(uint32_t), gw_name ## _param_0_too_big); \
+    UVISOR_STATIC_ASSERT(sizeof(p1_type) <= sizeof(uint32_t), gw_name ## _param_1_too_big); \
+    UVISOR_STATIC_ASSERT(sizeof(p2_type) <= sizeof(uint32_t), gw_name ## _param_2_too_big); \
+    UVISOR_STATIC_ASSERT(sizeof(p3_type) <= sizeof(uint32_t), gw_name ## _param_3_too_big);
+
+#define _UVISOR_BOX_RPC_GATEWAY_SYNC_CALLER(fn_name, ...) \
+    __UVISOR_BOX_MACRO(__VA_ARGS__, _UVISOR_BOX_RPC_GATEWAY_SYNC_CALLER_4, \
+                                    _UVISOR_BOX_RPC_GATEWAY_SYNC_CALLER_3, \
+                                    _UVISOR_BOX_RPC_GATEWAY_SYNC_CALLER_2, \
+                                    _UVISOR_BOX_RPC_GATEWAY_SYNC_CALLER_1, \
+                                    _UVISOR_BOX_RPC_GATEWAY_SYNC_CALLER_0)(fn_name, __VA_ARGS__)
+
+#define _UVISOR_BOX_RPC_GATEWAY_SYNC_CALLER_0(fn_name, ...) \
+    static uint32_t __sgw_sync_ ## fn_name(void) \
+    { \
+        TFN_Ptr fp = (TFN_Ptr) fn_name; \
+        return rpc_fncall(0, 0, 0, 0, fp); \
+    }
+
+#define _UVISOR_BOX_RPC_GATEWAY_SYNC_CALLER_1(fn_name, ...) \
+    static uint32_t __sgw_sync_ ## fn_name(uint32_t p0) \
+    { \
+        TFN_Ptr fp = (TFN_Ptr) fn_name; \
+        return rpc_fncall(p0, 0, 0, 0, fp); \
+    }
+
+#define _UVISOR_BOX_RPC_GATEWAY_SYNC_CALLER_2(fn_name, ...) \
+    static uint32_t __sgw_sync_ ## fn_name(uint32_t p0, uint32_t p1) \
+    { \
+        TFN_Ptr fp = (TFN_Ptr) fn_name; \
+        return rpc_fncall(p0, p1, 0, 0, fp); \
+    }
+
+#define _UVISOR_BOX_RPC_GATEWAY_SYNC_CALLER_3(fn_name, ...) \
+    static uint32_t __sgw_sync_ ## fn_name(uint32_t p0, uint32_t p1, uint32_t p2) \
+    { \
+        TFN_Ptr fp = (TFN_Ptr) fn_name; \
+        return rpc_fncall(p0, p1, p2, 0, fp); \
+    }
+
+#define _UVISOR_BOX_RPC_GATEWAY_SYNC_CALLER_4(fn_name, ...) \
+    static uint32_t __sgw_sync_ ## fn_name(uint32_t p0, uint32_t p1, uint32_t p2, uint32_t p3) \
+    { \
+        TFN_Ptr fp = (TFN_Ptr) fn_name; \
+        return rpc_fncall(p0, p1, p2, p3, fp); \
+    }
+
+#define _UVISOR_BOX_RPC_GATEWAY_ASYNC_CALLER(fn_name, ...) \
+    __UVISOR_BOX_MACRO(__VA_ARGS__, _UVISOR_BOX_RPC_GATEWAY_ASYNC_CALLER_4, \
```

Comment:
If this macro allows an empty `__VA_ARGS__` then the macro to use should actually be [this one](https://github.com/ARMmbed/uvisor/blob/master/api/inc/uvisor_exports.h#L62), as used [here](https://github.com/ARMmbed/uvisor/blob/master/api/inc/uvisor_exports.h#L71).

For a long time I've wanted to unify our variadic macros creation, as I'd like to remove duplications like this one (`__UVISOR_BOX_MACRO`).